db/migrate/20180806151925_add_subnet_name_unique_constraint.rb

```diff
@@ -0,0 +1,58 @@
+class AddSubnetNameUniqueConstraint < ActiveRecord::Migration[5.1]
+  def up
+    deduplicate_subnets
+    # we need to specify the lenght to make MySQL happy
+    add_index :subnets, :name, :unique => true, length: { name: 255 }
+  end
+
+  def down
+    remove_index :subnets, :name
+  end
+
+  private
+
+  def deduplicate_subnets
+    duplicate_names = Subnet.unscoped.select(:name).group(:name).having('count(name) > 1').pluck(:name)
+    Rails.logger.info("Found #{duplicate_names} duplicate subnet names.")
+    duplicate_names.each do |name|
+      suffix = 0
+      subnets = Subnet.unscoped.where(name: name)
+      subnet_groups = group_identic_subnets(subnets)
+      subnet_groups.each do |subnet_group|
+        if suffix == 0
+          new_name = name
+        else
+          new_name = "#{name}-#{suffix}"
+        end
+        suffix = suggest_unique_suffix(name, suffix)
+        merge_subnets(new_name, subnet_group)
+      end
+    end
+  end
+
+  def merge_subnets(new_name, subnets)
+    winner, *loosers = subnets
+    Rails.logger.info("Merging duplicate subnets with name '#{winner.name}', " \
+                        "keeping #{winner.id}, merging with #{loosers.map(&:id)}, "\
+                        "new name is '#{new_name}'")
+    Hostgroup.unscoped.where(:subnet_id => loosers.map(&:id)).update_all(:subnet_id => winner.id)
+    Nic::Base.unscoped.where(:subnet_id => loosers.map(&:id)).update_all(:subnet_id => winner.id)
+    SubnetParameter.unscoped.where(:reference_id => loosers.map(&:id)).update_all(:reference_id => winner.id)
+    Subnet.unscoped.where(id: loosers).destroy_all
```

Comment:
I'd use `destroy_all!` or something like that - we want to error out early during this process if deletion does not go well.

Reply:
https://stackoverflow.com/questions/32928476/are-there-destroy-all-delete-all-in-rails